File: go/src/kanzi/transform/BWTS.go
/*
Copyright 2011-2013 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package transform

import (
	"errors"
	"fmt"
	"kanzi"
	"kanzi/util"
)

// Bijective version of the Burrows-Wheeler Transform
// The main advantage over the regular BWT is that there is no need for a primary
// index (hence the bijectivity). BWTS is about 10% slower than BWT.
// Forward transform based on the code at https://code.google.com/p/mk-bwts/
// by Neal Burns and DivSufSort (port of libDivSufSort by Yuta Mori)

const (
	BWTS_MAX_BLOCK_SIZE = 1024 * 1024 * 1024 // 1 GB (30 bits)
)

type BWTS struct {
	buffer  []int
	buckets []int
	saAlgo  *util.DivSufSort
}

func NewBWTS() (*BWTS, error) {
	this := new(BWTS)
	this.buffer = make([]int, 0)
	this.buckets = make([]int, 256)
	return this, nil
}

func (this *BWTS) Forward(src, dst []byte, length uint) (uint, uint, error) {
	if src == nil {
		return 0, 0, errors.New("Input buffer cannot be null")
	}

	if dst == nil {
		return 0, 0, errors.New("Output buffer cannot be null")
	}

	if kanzi.SameByteSlices(src, dst, false) {
		return 0, 0, errors.New("Input and output buffers cannot be equal")
	}

	count := int(length)

	if count > maxBWTSBlockSize() {
		errMsg := fmt.Sprintf("Block size is %v, max value is %v", count, maxBWTSBlockSize())
		return 0, 0, errors.New(errMsg)
	}

	if count > len(src) {
		errMsg := fmt.Sprintf("Block size is %v, input buffer length is %v", count, len(src))
		return 0, 0, errors.New(errMsg)
	}

	if count < 2 {
		if count == 1 {
			dst[0] = src[0]
		}

		return uint(count), uint(count), nil
	}

	// Lazy dynamic memory allocations
	if len(this.buffer) < count {
		this.buffer = make([]int, count)
	}

	if this.saAlgo == nil {
		var err error

		if this.saAlgo, err = util.NewDivSufSort(); err != nil {
			return 0, 0, err
		}
	} else {
		this.saAlgo.Reset()
	}

	// Compute suffix array
	sa := this.saAlgo.ComputeSuffixArray(src[0:count])

	// Aliasing
	isa := this.buffer

	for i := 0; i < count; i++ {
		isa[sa[i]] = i
	}

	min := isa[0]
	idxMin := 0

	for i := 1; i < count && min > 0; i++ {
		if isa[i] >= min {
			continue
		}

		headRank := this.moveLyndonWordHead(sa, src, count, idxMin, i-idxMin, min)
		refRank := headRank

		for j := i - 1; j > idxMin; j-- {
			// iterate through the new lyndon word from end to start
			testRank := isa[j]
			startRank := testRank

			for testRank < count-1 {
				nextRankStart := sa[testRank+1]

				if j > nextRankStart || src[j] != src[nextRankStart] || refRank < isa[nextRankStart+1] {
					break
				}

				sa[testRank] = nextRankStart
				isa[nextRankStart] = testRank
				testRank++
			}

			sa[testRank] = j
			isa[j] = testRank
			refRank = testRank

			if startRank == testRank {
				break
			}
		}

		min = isa[i]
		idxMin = i
	}

	min = count

	for i := 0; i < count; i++ {
		if isa[i] >= min {
			dst[isa[i]] = src[i-1]
			continue
		}

		if min < count {
			dst[min] = src[i-1]
		}

		min = isa[i]
	}

	dst[0] = src[count-1]
	return uint(count), uint(count), nil
}

func (this *BWTS) moveLyndonWordHead(sa []int, data []byte, count, start, size, rank int) int {
	isa := this.buffer
	end := start + size

	for rank+1 < count {
		nextStart0 := sa[rank+1]

		if nextStart0 <= end {
			break
		}

		nextStart := nextStart0
		k := 0

		for k < size && nextStart < count && data[start+k] == data[nextStart] {
			k++
			nextStart++
		}

		if k == size && rank < isa[nextStart] {
			break
		}

		if k < size && nextStart < count && data[start+k] < data[nextStart] {
			break
		}

		sa[rank] = nextStart0
		isa[nextStart0] = rank
		rank++
	}

	sa[rank] = start
	isa[start] = rank
	return rank
}

func (this *BWTS) Inverse(src, dst []byte, length uint) (uint, uint, error) {
	if src == nil {
		return 0, 0, errors.New("Input buffer cannot be null")
	}

	if dst == nil {
		return 0, 0, errors.New("Output buffer cannot be null")
	}

	if kanzi.SameByteSlices(src, dst, false) {
		return 0, 0, errors.New("Input and output buffers cannot be equal")
	}

	count := int(length)

	if count > maxBWTSBlockSize() {
		errMsg := fmt.Sprintf("Block size is %v, max value is %v", length, maxBWTSBlockSize())
		return 0, 0, errors.New(errMsg)
	}

	if count < 2 {
		if count == 1 {
			dst[0] = src[0]
		}

		return uint(count), uint(count), nil
	}

	// Lazy dynamic memory allocation
	if len(this.buffer) < count {
		this.buffer = make([]int, count)
	}

	// Aliasing
	buckets_ := this.buckets
	lf := this.buffer

	// Initialize histogram
	for i := range this.buckets {
		buckets_[i] = 0
	}

	for i := 0; i < count; i++ {
		buckets_[src[i]]++
	}

	// Histogram
	for i, j := 0, 0; i < 256; i++ {
		t := buckets_[i]
		buckets_[i] = j
		j += t
	}

	for i := 0; i < count; i++ {
		lf[i] = buckets_[src[i]]
		buckets_[src[i]]++
	}

	// Build inverse
	for i, j := 0, count-1; j >= 0; i++ {
		if lf[i] < 0 {
			continue
		}

		p := i

		for {
			dst[j] = src[p]
			j--
			t := lf[p]
			lf[p] = -1
			p = t

			if lf[p] < 0 {
				break
			}
		}
	}

	return uint(count), uint(count), nil
}

func maxBWTSBlockSize() int {
	return BWTS_MAX_BLOCK_SIZE
}

File: java/src/main/java/kanzi/transform/BWTS.java
/*
Copyright 2011-2017 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package kanzi.transform;

import kanzi.ByteTransform;
import kanzi.SliceByteArray;



// Bijective version of the Burrows-Wheeler Transform
// The main advantage over the regular BWT is that there is no need for a primary
// index (hence the bijectivity). BWTS is about 10% slower than BWT.
// Forward transform based on the code at https://code.google.com/p/mk-bwts/ 
// by Neal Burns and DivSufSort (port of libDivSufSort by Yuta Mori)

public class BWTS implements ByteTransform
{
    private static final int MAX_BLOCK_SIZE = 1024*1024*1024; // 1 GB (30 bits)
 
    private int[] buffer1;
    private int[] buffer2;
    private final int[] buckets;
    private DivSufSort saAlgo;

    
    public BWTS()
    {
       this.buffer1 = new int[0];  
       this.buffer2 = new int[0];  
       this.buckets = new int[256];
    }


    // Not thread safe
    @Override
    public boolean forward(SliceByteArray src, SliceByteArray dst)
    {
        if ((!SliceByteArray.isValid(src)) || (!SliceByteArray.isValid(dst)))
           return false;

        if (src.array == dst.array)
           return false;
                
        final int count = src.length;
        
        if (count > maxBlockSize())
           return false;

        if (dst.index + count > dst.array.length)
           return false;
              
        final byte[] input = src.array;
        final byte[] output = dst.array;
        final int srcIdx = src.index;
        final int dstIdx = dst.index;

        if (count < 2)
        {
           if (count == 1)
              output[dst.index++] = input[src.index++];

           return true;
        }
        
        if (this.saAlgo == null)
           this.saAlgo = new DivSufSort();

        // Lazy dynamic memory allocations
        if (this.buffer1.length < count)
           this.buffer1 = new int[count];

        if (this.buffer2.length < count)
           this.buffer2 = new int[count];
                
        // Aliasing
        final int[] sa = this.buffer1;
        final int[] isa = this.buffer2;

        this.saAlgo.computeSuffixArray(input, sa, srcIdx, count);
        
        for (int i=0; i<count; i++)
           isa[sa[i]] = i;

        int min = isa[0];
        int idxMin = 0;
        
        for (int i=1; ((i<count) && (min>0)); i++) 
        {
            if (isa[i] >= min) 
               continue;
            
            int refRank = this.moveLyndonWordHead(sa, isa, input, count, srcIdx, idxMin, i-idxMin, min);

            for (int j=i-1; j>idxMin; j--) 
            { 
                // iterate through the new lyndon word from end to start
                int testRank = isa[j];
                int startRank = testRank;

                while (testRank < count-1) 
                {                 
                   int nextRankStart = sa[testRank+1];

                   if ((j > nextRankStart) || (input[srcIdx+j] != input[srcIdx+nextRankStart]) 
                            || (refRank < isa[nextRankStart+1])) 
                      break;

                   sa[testRank] = nextRankStart;
                   isa[nextRankStart] = testRank;
                   testRank++;
                }

                sa[testRank] = j;
                isa[j] = testRank;
                refRank = testRank;

                if (startRank == testRank) 
                   break;
            }

            min = isa[i];
            idxMin = i;
        }
        
        min = count;
        final int srcIdx2 = srcIdx - 1;
        
        for (int i=0; i<count; i++) 
        {
           if (isa[i] >= min) 
           {
              output[dstIdx+isa[i]] = input[srcIdx2+i];
              continue;
           }

           if (min < count)
              output[dstIdx+min] = input[srcIdx2+i];
           
           min = isa[i];
        }
   
        output[dstIdx] = input[srcIdx2+count];       
        src.index += count;
        dst.index += count; 
        return true;
    }

    
    private int moveLyndonWordHead(int[] sa, int[] isa, byte[] data, int count, int srcIdx, int start, int size, int rank)
    {
       final int end = start + size;
       final int startIdx = srcIdx + start;

       while (rank+1 < count)
       {
          final int nextStart0 = sa[rank+1];
          
          if (nextStart0 <= end)
             break;         
          
          int nextStart = nextStart0;
          int k = 0;

          while ((k < size) && (nextStart < count) && (data[startIdx+k] == data[srcIdx+nextStart]))
          {
              k++;
              nextStart++;
          }

          if ((k == size) && (rank < isa[nextStart]))
             break;

          if ((k < size) && (nextStart < count) && ((data[startIdx+k] & 0xFF) < (data[srcIdx+nextStart] & 0xFF)))
             break;         

          sa[rank] = nextStart0;
          isa[nextStart0] = rank;
          rank++;
       }
       
       sa[rank] = start;
       isa[start] = rank;
       return rank;
    }
    

    // Not thread safe
    @Override
    public boolean inverse(SliceByteArray src, SliceByteArray dst)
    {
        if ((!SliceByteArray.isValid(src)) || (!SliceByteArray.isValid(dst)))
           return false;

        if (src.array == dst.array)
           return false;
        
       final int count = src.length;
       
       if (count > maxBlockSize())
          return false;
       
       if (dst.index + count > dst.array.length)
          return false;
              
       if (count < 2)
       {
          if (count == 1)
             dst.array[dst.index++] = src.array[src.index++];

          return true;
       }
       
       final byte[] input = src.array;
       final byte[] output = dst.array;
       final int srcIdx = src.index;
       final int dstIdx = dst.index;
       
       // Lazy dynamic memory allocation
       if (this.buffer1.length < count)
          this.buffer1 = new int[count];

       // Aliasing
       final int[] buckets_ = this.buckets;
       final int[] lf = this.buffer1;
       
       // Initialize histogram
       for (int i=0; i<256; i++)
          buckets_[i] = 0;

       for (int i=0; i<count; i++)
          buckets_[input[srcIdx+i] & 0xFF]++;

       // Histogram
       for (int i=0, sum=0; i<256; i++)
       {
          sum += buckets_[i];
          buckets_[i] = sum - buckets_[i];
       }
       
       for (int i=0; i<count; i++)
          lf[i] = buckets_[input[srcIdx+i] & 0xFF]++;
       
       // Build inverse
       for (int i=0, j=dstIdx+count-1; j>=dstIdx; i++) 
       {         
          if (lf[i] < 0) 
             continue;  

          int p = i;
           
          do 
          {
             output[j] = input[srcIdx+p];
             j--;
             final int t = lf[p];
             lf[p] = -1;
             p = t;           
          } 
          while (lf[p] >= 0);
       }

       src.index += count;
       dst.index += count;
       return true;
    }
    
    
    private static int maxBlockSize() 
    {
       return MAX_BLOCK_SIZE;      
    }       
}
